src/mpl/src/gavl/mpl_gavl.c

```diff
@@ -205,71 +300,17 @@ int MPL_gavl_tree_insert(MPL_gavl_tree_t gavl_tree, const void *addr, uintptr_t
     } else {
-        DECLARE_STACK(gavl_tree_node_s *, node_stack);
+        DECLARE_STACK(node_stack);
         gavl_tree_node_s *cur_node = gavl_tree_iptr->root;
         int direction;
-        do {
-            int cmp_ret = gavl_subset_cmp_func((uintptr_t) node_ptr->addr, node_ptr->len, cur_node);
-            if (cmp_ret == SEARCH_LEFT) {
-                if (cur_node->left != NULL) {
-                    STACK_PUSH(node_stack, cur_node);
-                    cur_node = cur_node->left;
-                    continue;
-                } else {
-                    direction = SEARCH_LEFT;
-                }
-            } else if (cmp_ret == SEARCH_RIGHT) {
-                if (cur_node->right != NULL) {
-                    STACK_PUSH(node_stack, cur_node);
-                    cur_node = cur_node->right;
-                    continue;
-                } else {
-                    direction = SEARCH_RIGHT;
-                }
-            } else {
-                /* we cannot insert the duplicate buffer */
-                gavl_tree_iptr->gavl_free_fn((void *) node_ptr->val);
-                MPL_free(node_ptr);
-                goto fn_exit;
-            }
-
-            if (direction == SEARCH_LEFT)
-                cur_node->left = node_ptr;
-            else
-                cur_node->right = node_ptr;
-            node_ptr->parent = cur_node;
-
-          stack_recovery:
-            gavl_update_node_info(cur_node);
-
-            int lheight = cur_node->left == NULL ? 0 : cur_node->left->height;
-            int rheight = cur_node->right == NULL ? 0 : cur_node->right->height;
-            if (lheight - rheight > 1) {
-                gavl_tree_node_s *lnode = cur_node->left;
-                int llheight = lnode->left == NULL ? 0 : lnode->left->height;
-                if (llheight + 1 == lheight)
-                    gavl_right_rotation(cur_node, lnode);
-                else
-                    gavl_left_right_rotation(cur_node, lnode);
-            } else if (rheight - lheight > 1) {
-                gavl_tree_node_s *rnode = cur_node->right;
-                int rlheight = rnode->left == NULL ? 0 : rnode->left->height;
-                if (rlheight + 1 == rheight)
-                    gavl_right_left_rotation(cur_node, rnode);
-                else
-                    gavl_left_rotation(cur_node, rnode);
-            }
-
-            if (!STACK_EMPTY(node_stack)) {
-                STACK_POP(node_stack, cur_node);
-                goto stack_recovery;
-            } else {
-                break;
-            }
-        } while (1);
+        gavl_tree_insert_search(gavl_tree_iptr, node_ptr, cur_node, &direction, &cur_node,
+                                node_stack);
+        if (direction == NO_BUFFER_MATCH)
+            goto fn_exit;
 
-        while (gavl_tree_iptr->root->parent != NULL)
-            gavl_tree_iptr->root = gavl_tree_iptr->root->parent;
+        gavl_tree_insert_node(direction, node_ptr, cur_node);
```

Comment:
Expand this. It is a part of three-way switch: insert_left, insert_right, and break. Hiding them breaks the integrity of the switch and makes the code more complex.